src/state_circuit/memory.rs

```diff
@@ -129,33 +133,53 @@ impl<F: FieldExt, const NUM_STEPS: usize> Config<F, NUM_STEPS> {
         // Constraints for each step. These are activated when q_memory is nonzero.
         meta.create_gate("Memory operation", |meta| {
             let q_memory = meta.query_selector(q_memory);
+            let q_memory_init = meta.query_selector(q_memory_init);
```

Comment:
We want to define a synthetic `q_memory_init` selector within the gate, instead of "hard-coding" the intervals at which we enable the init selector outside the gate. This is because each memory location sees an arbitrary number of operations per proof, and we cannot know ahead of time the offsets at which we check for an init.
```suggestion
            // If address_cur != address_prev, this is an `init`.
            let q_memory_init = {
                let address_prev = meta.query_advice(address, Rotation::prev());
                let address_cur = meta.query_advice(address, Rotation::cur());    
                address_cur - address_prev
            };
```